frontend/model/contracts/group.js

```diff
@@ -227,6 +227,158 @@ DefineContract({
+
+          if (amount <= 0) { continue }
+
+          const partialAmount = existingAmount - amount
+          const existingPayment = {}
+          if (partialAmount > 0) {
+            // TODO/BUG this only work if the payment is done in 2 parts. if done in >=3 won't work.
+            const receivePartial = received.find((r) => r.username === p.to && r.amount === partialAmount)
+            if (receivePartial) {
+              existingPayment.hash = receivePartial.hash
+            }
+          }
+          payments.push({
+            ...existingPayment,
+            ...p,
+            total: existingAmount,
+            partial: partialAmount > 0
+          })
+        }
+        return payments
+      })()
+
+      return { sent, todo, received, toBeReceived }
+    },
+    ourPaymentsSummary (state, getters) {
+      const { todo, sent, toBeReceived, received } = getters.ourPayments
+      const isCompleted = (p) => p.data.status === PAYMENT_COMPLETED
+      const isPartialCount = (list) => list.filter(p => p.partial).length
+      const getUniqPayments = (users) => {
+        // We need to filter the partial payments already done (sent/received).
+        // E.G. We need to send 4 payments. We've sent 1 full payment and another
+        // in 2 parts. The total must be 2, instead of 3. A quick way to solve this
+        // is by listing all usernames we sent to and count the uniq ones.
+        return uniq(users).length
+      }
+
+      if (getters.ourGroupProfile.incomeDetailsType === 'incomeAmount') {
+        const receivedCompleted = received.filter(isCompleted)
+        const pPartials = isPartialCount(toBeReceived)
+        const pByUser = {
+          toBeReceived: toBeReceived.map(p => p.from),
+          received: received.map(p => p.meta.username)
+        }
+        return {
+          paymentsDone: getUniqPayments(pByUser.received) - pPartials,
+          hasPartials: pPartials > 0,
+          paymentsTotal: getUniqPayments([...pByUser.toBeReceived, ...pByUser.received]),
+          amountDone: receivedCompleted.reduce((total, p) => total + p.data.amount, 0),
+          amountTotal: toBeReceived.reduce((total, p) => total + p.amount, 0) + received.reduce((total, p) => total + p.data.amount, 0)
+        }
+      } else {
+        const sentCompleted = sent.filter(isCompleted)
+        const pPartials = isPartialCount(todo)
+        const pByUser = {
+          todo: todo.map(p => p.to),
+          sent: sent.map(p => p.data.toUser)
+        }
+        return {
+          paymentsDone: getUniqPayments(pByUser.sent) - pPartials,
+          hasPartials: pPartials > 0,
+          paymentsTotal: getUniqPayments([...pByUser.todo, ...pByUser.sent]),
+          amountDone: sentCompleted.reduce((total, p) => total + p.data.amount, 0),
+          amountTotal: todo.reduce((total, p) => total + p.amount, 0) + sent.reduce((total, p) => total + p.data.amount, 0)
+        }
+      }
+    },
```

Comment:
@sandrina-p since both of these getters rely on data that does not come directly from the contract messages / current-contract-state (e.g. `currentMonthstamp()`), please move both getters into `state.js`

Reply:
Following that mindset, other getters should also be moved, such as `groupIncomeDistribution`, `groupIncomeAdjustedDistribution` and `thisMonthsPaymentInfo`. Should I move all of them too?

Reply:
@sandrina-p Yep! Excellent catch! And maybe you could add a comment about this above all of them (in `state.js`) explaining all of this (e.g. why they are there are not in group.js)?

Reply:
@taoeffect could you explain again what's the problem with having `currentMonthstamp` at `group.js` and why is it okay to be at `state.js`? My memory is failing me here...

Also, I wonder if having `currentMonthstamp` in the getters isn't dangerous because it's not a "static value", but rather a dynamic value that can change over time without the user action. This may conflict with some contract message `process`... Imagine something related to the current month happens on 30 June and then another member logs one day after (1 July) to process the message from the previous month. A getter result won't be the same because of `currentMonthstamp`. I think there isn't any scenario yet that does this, but it's something we need to take attention.

Reply:
> Also, I wonder if having currentMonthstamp in the getters isn't dangerous because it's not a "static value", but rather a dynamic value that can change over time without the user action

It's dangerous precisely because of that, and that is why it must not be inside of the contract getters but only in the Vuex getters.

The getters in state.js are Vuex-only getters. These are getters that are not used to build up the contract state, which can only be built-up using the information that's coming from the messages being sent to the contract.

Since `currentMonthstamp` is, as you point out, a dynamic value that comes *not* directly from the data in contract messages, but from the end-user's computer, and it can only be used in the Vuex-getters that are used to build up the ephemeral, non-contract state that is in the UI.

The *contract getters* can only access information from two places:

1. each `GIMessage` that comes in (e.g. that's passed to the `process` function)
2. the existing *contract state* (and **not** any other state, e.g. no Vuex-state is allowed to be used that exists outside of the contract state)

Reply:
gotcha, so we are on the same page. moved everything to `state.js` and created a new file with `groupIncomeDistribution` to be imported from both files (`state.js` and `group.js`)